### PR TITLE
fix(curriculum): new step for empty title bug

### LINKED ISSUE #54588: User can create a task with the white space/blank title - todoApp
### Describe the Issue

***STEPS TO REPRODUCE:***
1. Start todo App
2. Click on "Add New Task" button
3. On the "Title" field click on the "Space" key (long key on your keyboard) to create a blank space
4. Click on "Add Task"

***ACTUAL RESULT:***
The user created empty task even though title field is required

### Affected Page

https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures-v8/learn-localstorage-by-building-a-todo-app/step-65

### Your code

```JavaScript
taskForm.addEventListener('submit', (e) => {
    e.preventDefault();
    addOrUpdateTask();
});
```

### Expected behavior

User should not be able to create an empty title, as it goes against what we set on HTML. 
```HTML
<input required type="text" class="form-control" id="title-input" value="" />
```
_________________________________________________________________________
as discussed with @jdwilkin4 under the issue https://github.com/freeCodeCamp/freeCodeCamp/issues/54524

_________________________________________________________________________
Ex. of expected behavior 1:

```JavaScript
const isTitleEmpty = (val) => !val.trim().length


taskForm.addEventListener("submit", (e) => {
  e.preventDefault();

  if(isTitleEmpty(titleInput.value)){
    alert("Please provide a title")
    return
  }
  
  addOrUpdateTask();
});
````

Ex. of expected behavior 2:
***JS***
```JavaScript
const errorMessage = document.getElementById("errorMessage");
const isEmpty = (str) => !str || !str.trim();

taskForm.addEventListener('submit', (e) => {
    e.preventDefault();
    if (!isEmpty(titleInput.value)) {
        addOrUpdateTask();
    } else {
        // Create an error message element
        errorMessage.textContent = "Error: title cannot be an empty value!";
        errorMessage.style.display = "block";

        // Remove the error message after 3 seconds
        setTimeout(() => {
            errorMessage.style.display = "none";
        }, 3000);
    }
});
```
***HTML***
```HTML
<div class="task-form-body">
          <label class="task-form-label" for="title-input">Title</label>
          <input required type="text" class="form-control" id="title-input" value="" />
          <label id="errorMessage" style="display: none; color: red; font-weight: bold;"></label>
          <label class="task-form-label" for="date-input">Date</label>
          <input type="date" class="form-control" id="date-input" value="" />
          <label class="task-form-label" for="description-input">Description</label>
          <textarea class="form-control" id="description-input" cols="30" rows="5"></textarea>
</div>
```

### Screenshots

Screenshot 2024-04-30 141823

### System

Device: Laptop
OS: Windows 11 Home
Browser: Chrome
Version: 124.0.6367.61

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by gagan-bhullar-tech**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #54588

### CODE REVIEW

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-localstorage-by-building-a-todo-app/66ad0f178ed5791ed898fe56.md [hunk 1]
```diff
@@ -0,0 +1,794 @@
+---
+id: 66ad0f178ed5791ed898fe56
+title: Step 66
+challengeType: 0
+dashedName: step-66
+---
+
+# --description--
+
+If you try to add a new task, add space in title field and then click on the `Add New Task` button, you will notice a bug.
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-localstorage-by-building-a-todo-app/66ad0f178ed5791ed898fe56.md:10]:
```suggestion
If you try to add a new task, add a single space in title field, and click on the `Add New Task` button, you will notice a bug.
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-localstorage-by-building-a-todo-app/66ad0f178ed5791ed898fe56.md [hunk 2]
```diff
@@ -0,0 +1,794 @@
...
+
+Empty task will be created even though title field is required. To fix this, you need add an `if` condition checking for empty title field value.
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-localstorage-by-building-a-todo-app/66ad0f178ed5791ed898fe56.md:12]:
```suggestion
An empty task will be created even though title field is required. To fix this, you need add an `if` condition checking for a value that would appear empty.
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-localstorage-by-building-a-todo-app/66ad0f178ed5791ed898fe56.md [hunk 3]
```diff
@@ -0,0 +1,794 @@
...
+
+```js
+assert.match(code, /if\s*\(\s*!titleInput\.value\.trim\(\s*\)\s*\)\s*/)
+```
+
+You should call the `alert("Please provide a title")` in your `if` block.
+
+```js
+console.log("here is")
+console.log(code)
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-localstorage-by-building-a-todo-app/66ad0f178ed5791ed898fe56.md:29]:
```suggestion
```

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-localstorage-by-building-a-todo-app/66ad0f178ed5791ed898fe56.md [hunk 4]
```diff
@@ -0,0 +1,792 @@
...
+title: Step 66
+challengeType: 0
+dashedName: step-66
+---
+
+# --description--
+
+If you try to add a new task, add a single space in title field, and click on the `Add New Task` button, you will notice a bug.
+
+An empty task will be created even though title field is required. To fix this, you need add an `if` condition checking for a value that would appear empty.
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-localstorage-by-building-a-todo-app/66ad0f178ed5791ed898fe56.md:12]:
```suggestion
An empty task will be created even though the title field is required. To fix this, you need to add an `if` condition checking for a value that would appear empty.
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-localstorage-by-building-a-todo-app/66ad0f178ed5791ed898fe56.md [hunk 5]
```diff
@@ -0,0 +1,792 @@
+---
+id: 66ad0f178ed5791ed898fe56
+title: Step 66
+challengeType: 0
+dashedName: step-66
+---
+
+# --description--
+
+If you try to add a new task, add a single space in title field, and click on the `Add New Task` button, you will notice a bug.
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-localstorage-by-building-a-todo-app/66ad0f178ed5791ed898fe56.md:10]:
```suggestion
If you try to add a new task, write single space in title field, and click on the `Add New Task` button, you will notice a bug.
```

**Review by majestic-owl448 — CHANGES_REQUESTED** [review]
Can you please adjust these sentences

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-localstorage-by-building-a-todo-app/66ad0f178ed5791ed898fe56.md (hunk 4 above)
Comment (gagan-bhullar-tech) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-localstorage-by-building-a-todo-app/66ad0f178ed5791ed898fe56.md:12]: review comments

**Review by majestic-owl448 — APPROVED** [review]
LGTM 👍

**Review by moT01 — COMMENTED** [review]
I don't think the description explains enough for users to be able to complete this. The only way they will know what code to write is by looking at the hints. If the test is going to require using certain code or methods, we need to explain that in the description.

**Comment by gagan-bhullar-tech** [comment]
@moT01 can you please let me know what should be the description in that case.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-localstorage-by-building-a-todo-app/66ad0f178ed5791ed898fe56.md [hunk 6]
```diff
@@ -0,0 +1,792 @@
...
+title: Step 66
+challengeType: 0
+dashedName: step-66
+---
+
+# --description--
+
+If you try to add a new task, write single space in title field, and click on the `Add New Task` button, you will notice a bug.
+
+An empty task will be created even though the title field is required. To fix this, you need to add an `if` condition checking for a value that would appear empty.
```
Comment (moT01) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-localstorage-by-building-a-todo-app/66ad0f178ed5791ed898fe56.md:12]:
```suggestion
Finally, there's one bug to fix. If you create a task with only spaces for the title, an empty task will be created even though the title is required.

Fix that by adding a check at the beginning of the function for if `!titleInput.value.trim()`. If there's no title, show an alert with the text `Please provide a title` and `return` after that.
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-localstorage-by-building-a-todo-app/66ad0f178ed5791ed898fe56.md [hunk 7]
```diff
@@ -0,0 +1,792 @@
...
+You should add an `if` statement with the condition `!titleInput.value.trim()` inside `addOrUpdateTask` function.
+
+```js
+assert.match(code, /if\s*\(\s*!titleInput\.value\.trim\(\s*\)\s*\)\s*/)
+```
+
+You should call the `alert("Please provide a title")` in your `if` block.
+
+```js
+assert.match(code, /if\s*\(\s*!titleInput\.value\.trim\(\s*\)\s*\)\s*\{\s*\s*alert\s*\(\s*('|"|`)Please provide a title\1\s*\)\s*/)
```
Comment (moT01) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-localstorage-by-building-a-todo-app/66ad0f178ed5791ed898fe56.md:28]:
```suggestion
assert.match(code, /if\s*\(\s*!titleInput\.value\.trim\(\s*\)\s*\)\s*\{\s*alert\s*\(\s*('|"|`)Please provide a title\1\s*\)\s*/)
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-localstorage-by-building-a-todo-app/66ad0f178ed5791ed898fe56.md [hunk 8]
```diff
@@ -0,0 +1,792 @@
...
+```
+
+Your `if` statement should return from the function.
+
+```js
+assert.match(code, /if\s*\(\s*!titleInput\.value\.trim\(\s*\)\s*\)\s*\{\s*\s*alert\s*\(\s*('|"|`)Please provide a title\1\s*\)\s*;?\s*return\s*;?\s*\}/);
```
Comment (moT01) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-localstorage-by-building-a-todo-app/66ad0f178ed5791ed898fe56.md:34]:
```suggestion
assert.match(code, /if\s*\(\s*!titleInput\.value\.trim\(\s*\)\s*\)\s*\{\s*alert\s*\(\s*('|"|`)Please provide a title\1\s*\)\s*;?\s*return\s*;?\s*\}/);
```

**Review by moT01 — CHANGES_REQUESTED** [review]
Sorry for the delay @gagan-bhullar-tech. I left a comment with a new description. Let me know if you have any suggestions for improvements on it. Also found a few minor things that weren't needed in the regex tests.

**Comment by gagan-bhullar-tech** [comment]
@moT01 thanks. Updated PR.

**Review by moT01 — APPROVED** [review]

**Review by naomi-lgbt — APPROVED** [review]
Apologies for leaving this blocked.

**Pull request merged**